percona/controllers/pgc/pgc.go

```diff
@@ -266,71 +284,90 @@ func (c *Controller) getOperatorNamespaceList() ([]string, error) {
+		perconaPGClusters, err := c.Client.CrunchydataV1().PerconaPGClusters(namespace).List(ctx, metav1.ListOptions{})
 		if err != nil && !kerrors.IsNotFound(err) {
 			return errors.Wrap(err, "list perconapgclusters")
 		} else if err != nil {
 			// there is no perconapgclusters, so no need to continue
 			continue
 		}
-		for _, p := range perconaPGClusters.Items {
-			err = c.reconcileUsers(&p)
-			if err != nil && !kerrors.IsNotFound(err) {
-				return errors.Wrap(err, "reconcile users")
-			}
-			pgCluster, err := c.Client.CrunchydataV1().Pgclusters(ns).Get(ctx, p.Name, metav1.GetOptions{})
-			if err != nil && !kerrors.IsNotFound(err) {
-				return errors.Wrap(err, "get pgCluster")
+		for _, pi := range perconaPGClusters.Items {
+			nn := types.NamespacedName{
+				Name:      pi.Name,
+				Namespace: pi.Namespace,
 			}
+			l := c.lockers.LoadOrCreate(nn.String())
+			l.statusMutex.Lock()
 
-			pgClusterStatus := crv1.PgclusterStatus{}
-			replStatuses := make(map[string]crv1.PgreplicaStatus)
-			if pgCluster != nil {
-				pgClusterStatus = pgCluster.Status
-			}
-
-			selector := config.LABEL_PG_CLUSTER + "=" + p.Name
-			pgReplicas, err := c.Client.CrunchydataV1().Pgreplicas(ns).List(ctx, metav1.ListOptions{LabelSelector: selector})
+			p, err := c.Client.CrunchydataV1().PerconaPGClusters(pi.Namespace).Get(ctx, pi.Name, metav1.GetOptions{})
 			if err != nil && !kerrors.IsNotFound(err) {
-				return errors.Wrap(err, "get pgReplicas list")
-			}
-			if pgReplicas != nil {
-				for _, repl := range pgReplicas.Items {
-					replStatuses[repl.Name] = repl.Status
-				}
+				return errors.Wrap(err, "get perconapgluster")
 			}
 
-			if reflect.DeepEqual(p.Status.PGCluster, pgCluster.Status) && reflect.DeepEqual(p.Status.PGReplicas, replStatuses) {
-				return nil
-			}
-
-			value := crv1.PerconaPGClusterStatus{
-				PGCluster:  pgClusterStatus,
-				PGReplicas: replStatuses,
-			}
-
-			patch, err := kubeapi.NewJSONPatch().Replace("status")(value).Bytes()
+			err = c.reconcileUsers(p)
 			if err != nil {
-				return errors.Wrap(err, "create patch bytes")
+				log.Error(errors.Wrap(err, "reconcile users"))
 			}
-			_, err = c.Client.CrunchydataV1().PerconaPGClusters(p.Namespace).
-				Patch(ctx, p.Name, types.JSONPatchType, patch, metav1.PatchOptions{})
+			err = c.reconcileStatus(p, l.statusMutex)
 			if err != nil {
-				return errors.Wrap(err, "patch percona status")
+				log.Error(errors.Wrap(err, "reconcile status"))
 			}
 		}
 	}
 
 	return nil
 }
 
+func (c *Controller) reconcileStatus(cluster *crv1.PerconaPGCluster, statusMutex *sync.Mutex) error {
+	defer statusMutex.Unlock()
+
+	ctx := context.TODO()
+	pgCluster, err := c.Client.CrunchydataV1().Pgclusters(cluster.Namespace).Get(ctx, cluster.Name, metav1.GetOptions{})
+	if err != nil && !kerrors.IsNotFound(err) {
+		return errors.Wrap(err, "get pgCluster")
+	}
+
+	pgClusterStatus := crv1.PgclusterStatus{}
+	replStatuses := make(map[string]crv1.PgreplicaStatus)
+	if pgCluster != nil {
```

Comment:
is it possible `pgCluster` to be `nil`? If it's possible should we return earlier?

Reply:
https://github.com/percona/percona-postgresql-operator/pull/181/commits/ed0fd46c2efa1c89d88cd96a34961a807fa40913